Adjust "account not in good standing" display thresholds

002love opened the pull request:
Resolves #12533

This PR implements logic to hide the "bad standing" warning for minor infractions, as discussed in the issue

Key changes:
- warning is hidden for regular users if recent silences are <24h and inactive
- moderators still see the warning if any history exists (preserving visibility for staff)
- threshold of 3+ infringements triggers the warning regardless of length

Includes PHPUnit tests covering all logic branches

LiquidPL commented on app/Models/User.php:
```diff
@@ -1075,6 +1075,43 @@ public function isModerator()
+     * @param User|null $viewer The user viewing the profile.
+     * @return bool
+     */
+    public function hasBadStanding(?User $viewer = null): bool
```
change the name to `inBadStanding` to be more in line with wording on the website

LiquidPL commented on app/Models/User.php:
```diff
@@ -1075,6 +1075,43 @@ public function isModerator()
+        }
+
+        foreach ($silences as $silence) {
+            if ($silence->endTime()->isAfter()) {
```
`isAfter()` should have the timestamp passed into it

002love replied:
fixed

002love commented on tests/Models/UserTest.php:
```diff
@@ -269,4 +270,94 @@ public function testValidDiscordUsername(string $username, bool $valid)
+
+    public function testHasBadStandingForModeratorWithHistory()
+    {
+        config_set('osu.user.ban_persist_days', 28);
```
fixed

002love commented on app/Models/User.php:
```diff
@@ -1075,6 +1075,43 @@ public function isModerator()
+
+        if ($viewer?->isModerator()) {
+            return true;
+        }
```
fixed

002love commented on tests/Models/UserTest.php:
```diff
@@ -269,4 +270,94 @@ public function testValidDiscordUsername(string $username, bool $valid)
+
+        // no history
+        $this->assertFalse($user->hasBadStanding($viewer));
+    }
```
fixed

002love commented:
Thanks for the feedback, applied all requested changes:
- renamed `hasBadStanding` to `inBadStanding` in the model and moved the moderator logic to the UserTransformer 
- refactored `isAfter()` to `isFuture()`
- cleaned up `config_set` calls in unit tests

Userlog2 commented:
Thanks for taking the time to implement my idea!

One minor point of criticism, though: It's redundant (and in my opinion, slightly counterproductive) to show the "bad standing" warning if a user has an active silence because:

1. There is another equally obvious info message showing that the user is silenced,
2. The user *is silenced* and feeling the full effects of it; an extra red text isn't going to make it much more noticeable when they already can't chat, comment, etc.
3. It makes the purpose of the "bad standing" warning less clear. It's intended for severe/repeat offenders, so flashing it at a minor chat offender for 5 minutes could be a bit confusing.

Userlog2 commented on app/Models/User.php:
```diff
@@ -1075,6 +1075,38 @@ public function isModerator()
+            return false;
+        }
+
+        foreach ($silences as $silence) {
```
If I'm interpreting this code properly, this makes the warning *never* be shown if the user has an active silence. I meant that the "bad standing" warning should be independent of whether the silence is active or not.

Also, you may want to update the "key changes" list in the main page.

LiquidPL replied:
Yeah, I just realized that an active silence will always hide the warning, even if it's length would normally make it visible.

nanaya commented on app/Models/User.php:
```diff
@@ -1075,6 +1075,38 @@ public function isModerator()
+     *
+     * @return bool
+     */
+    public function inBadStanding(): bool
```
this should be done in the frontend instead of doing redundant query (which happens on all page access due to inclusion in transformer for current user json)

nanaya commented on resources/js/profile-page/account-standing.tsx:
```diff
@@ -78,6 +79,31 @@ const content: Record<Column, (props: ColumnProps) => JSX.Element | null> = {
+
+    if (core.currentUser?.is_moderator && history.length > 0) {
+      return true;
+    }
```
pretty sure this should check for silence, not entire history?

nanaya commented on resources/js/profile-page/account-standing.tsx:
```diff
@@ -78,6 +79,31 @@ const content: Record<Column, (props: ColumnProps) => JSX.Element | null> = {
+    const recentCount = silences.length;
+    const longestPeriod = Math.max(...silences.map((s) => s.length));
+
+    if (longestPeriod >= 2560 * 60 || recentCount >= 3) {
```
where did this 2560 come from? seems like a weird number to have (in term of time).

And just return the check instead of doing the equivalent of `if (condition) return true; return false`

Userlog2 replied:
2560 minutes = 42 hours 40 minutes ≈ 43 hours, the length of a "standard" long silence. Silence lengths are calculated with a simple exponential function, which in this case is 5min * 2⁹.

nanaya replied:
that should be put in the comment

002love commented on resources/js/profile-page/account-standing.tsx:
```diff
@@ -78,6 +79,27 @@ const content: Record<Column, (props: ColumnProps) => JSX.Element | null> = {
+    const history = user.account_history;
+    const silences = history.filter((d) => d.type === 'silence');
+
+    if (core.currentUser?.is_moderator && silences.length > 0) {
```
fixed

002love commented on resources/js/profile-page/account-standing.tsx:
```diff
@@ -6,6 +6,7 @@ import StringWithComponent from 'components/string-with-component';
 import TimeWithTooltip from 'components/time-with-tooltip';
 import UserLink from 'components/user-link';
 import UserAccountHistoryJson from 'interfaces/user-account-history-json';
+import core from 'osu-core-singleton';
```
fixed
